std/hash/poseidon_bn254.asm

```diff
@@ -52,70 +54,51 @@ machine PoseidonBN254(LASTBLOCK, operation_id) {
+    let C = [C_0, C_1, C_2];
 
-    // State of the Poseidon permutation
-    pol commit in0, in1, cap;
+    // State of the Poseidon permutation (2 rate elements and 1 capacity element)
+    pol commit state[STATE_SIZE];
 
     // The initial state of the Poseidon permutation
-    // (constrained to be equal to (in0, in1, cap) in the first row and then repeated until
+    // (constrained to be equal to state in the first row and then repeated until
     // the end of the block)
-    pol commit input_in0, input_in1, input_cap;
+    pol commit input_state[STATE_SIZE];
 
     // Add round constants
-    pol commit a0, a1, a2;
-    a0 = in0 + C_0;
-    a1 = in1 + C_1;
-    a2 = cap + C_2;
+    pol commit a[STATE_SIZE];
+    array::zip3(a, state, C, |a, state, C| a = state + C);
 
     // Compute S-Boxes (x^5)
-    pol commit x2_0, x4_0, x5_0;
-    x2_0 = a0 * a0;
-    x4_0 = x2_0 * x2_0;
-    x5_0 = x4_0 * a0;
-
-    pol commit x2_1, x4_1, x5_1;
-    x2_1 = a1 * a1;
-    x4_1 = x2_1 * x2_1;
-    x5_1 = x4_1 * a1;
-
-    pol commit x2_2, x4_2, x5_2;
-    x2_2 = a2 * a2;
-    x4_2 = x2_2 * x2_2;
-    x5_2 = x4_2 * a2;
+    pol commit x2[STATE_SIZE], x4[STATE_SIZE], x5[STATE_SIZE];
+    array::zip2(x2, a, |x2, a| x2 = a * a);
```

Comment:
By the way, I think we are not too far away from custom implementations for binary operators...

Reply:
How would that be used here? So you can define element-wise operations on arrays and write the whole thing as `x2 = a * a;`?